components/list/list-item-drag-mixin.js

```diff
@@ -33,86 +95,299 @@ export const ListItemDragMixin = superclass => class extends superclass {
+			case dragActions.up:
+				destinationKey = this.previousElementSibling && this.previousElementSibling.key;
+				this._moveItem(this.key, destinationKey, true);
+				break;
+			case dragActions.down:
+				destinationKey = this.nextElementSibling && this.nextElementSibling.key;
+				this._moveItem(this.key, destinationKey);
+				break;
+			case dragActions.first:
+				while (this.previousElementSibling) {
+					destinationKey = this.previousElementSibling;
+				}
+				this._moveItem(this.key, destinationKey, true);
+				break;
+			case dragActions.last:
+				while (this.nextElementSibling) {
+					destinationKey = this.nextElementSibling;
+				}
+				destinationKey = destinationKey && destinationKey.key;
+				this._moveItem(this.key, destinationKey);
+				break;
+			default:
+				break;
+		}
+
+	}
+
+	_dragLowerEnter(e) {
+		e.dataTransfer.dropEffect = 'move';
+		if (this._inBottomArea) {
+			const dropSpots = dropSpotsFactory();
+			dropSpots.setDestination(this, move.above);
+			this._inBottomArea = false;
+		}
+	}
+
+	_dragOver(e) {
+		const dropSpots = dropSpotsFactory();
+		dropSpots.updateTime(e.timeStamp);
+		e.preventDefault();
+	}
+
+	_dragStart(e) {
+		e.dataTransfer.setData('text/plain', `${this.dragText}`);
+		e.dataTransfer.effectAllowed = 'move';
+
+		const nodeImage = this.shadowRoot.querySelector('.d2l-list-item-drag-image') || this;
+		e.dataTransfer.setDragImage(nodeImage, 50, 50);
+
+		dropSpotsFactory(this);
+
+		setTimeout(() => {
+			this.beingDragged = true;
+		});
+	}
+
+	_dragTopEnter(e) {
+		e.dataTransfer.dropEffect = 'move';
+		const dropSpots = dropSpotsFactory();
+		dropSpots.setDestination(this, move.above);
+		this._inTopArea = true;
+	}
+
+	_dragUpperEnter(e) {
+		e.dataTransfer.dropEffect = 'move';
+		if (this._inTopArea) {
+			const dropSpots = dropSpotsFactory();
+			dropSpots.setDestination(this, move.below);
+			this._inTopArea = false;
+		}
+	}
+
+	_drop() {
+		const dropSpots = dropSpotsFactory();
+		dropSpots.setDestination(this, dropSpots.moveBeforeDestination);
+	}
+
+	_hostDragEnter(e) {
+		const dropSpots = dropSpotsFactory();
+		if (this === dropSpots.target) {
+			return;
+		}
+		dropSpots.addVisitor(this);
+		this._dropTarget = true;
+		e.dataTransfer.dropEffect = 'move';
+	}
+
+	_moveItem(targetKey, destinationKey, moveBeforeDestination = false, temporaryMovement = false) {
 		this.dispatchEvent(new CustomEvent('d2l-list-item-position', {
-			detail: detail,
+			detail: new NewPositionEventDetails({targetKey, destinationKey, moveBeforeDestination, temporaryMovement}),
 			bubbles: true
 		}));
 	}
 
-	_dragEnter() {
-		this._dragStartHandler();
+	_renderBottomPlacementMarker(renderTemplate) {
+		return this._bottomPlacementMarker ? html`<div class="d2l-list-item-drag-bottom-marker">${renderTemplate}</div>` : null;
+	}
+
+	_renderDraggableArea(templateMethod) {
+		templateMethod = templateMethod || (dragArea => dragArea);
+		return this.draggable && !this._keyboardMode ? templateMethod(html`
```

Comment:
`this.draggable && !this._keyboardMode` see more complicated cases on when to show and hide the drag items.